### PR TITLE
[i18n] Merge in changes from the i18n branch to master

### PR DESCRIPTION
**Pull request by beets**
Squashed commit of the following:

commit 4809a3b5fbd4b82b52f7104bf1e63cd98e37c158
Author: beets <beets@users.noreply.github.com>
Date:   Tue Jan 26 11:52:58 2021 -0800

    [i18n] fix server-side locale selection (#787)

    * fix server-side locale selection

    - only use locales that we have translations for - share the same logic with i18n name selection
    - remove extra lookups for names in the i18n path
    - try to resolve zh

    * add test for landing page child name i18n

commit c911b12010646c55ba8161565b6060ef941f5ef0
Author: beets <beets@users.noreply.github.com>
Date:   Fri Jan 22 16:12:25 2021 -0800

    [i18n] Propagate hl parameters (#781)

    - append hl parameter to all flask links that use url_for
    - append hl parameter to relevant js links (added a new react functional component for this)
    in both cases, only propagates hl parameter if it's not 'en'.

    - also disables 'explore more' link for non-en languages, since it won't be translated

commit 2643b33076288dfcf76ae70ace0e478888b513dd
Author: beets <beets@users.noreply.github.com>
Date:   Fri Jan 22 14:53:51 2021 -0800

    [i18n] Bundle in remaining strings (#778)

    - Bundle in messages for stats_var_labels and stats_var_titles (for en only, while translations are underway). This is quite manual at the moment - each module needs to specify up front which message bundle is required for that app so we only add required translations to each package.
    - Use server-specified locale on the client
    - Update bundle for new server-side strings
    - Submit compiled messages, since they add time to builds and don't change often
    - Update tests / cleanup code.

commit c9a2aea84953d066e1a232d9fafb5fa51437af7f
Author: beets <beets@users.noreply.github.com>
Date:   Tue Jan 19 11:25:51 2021 -0800

    [i18n] Prepare stat var titles for translation (#775)

    - Add a new json for stat var title messages, including descriptions. As discussed in group thread, formatting for some titles were updated, as well as many bugs fixed
    - Also fixed issues with other newly extracted messages while preparing import to TC

commit a2a684079d92bc74cebf86524675827a7e1f05fc
Author: beets <beets@users.noreply.github.com>
Date:   Fri Jan 15 09:18:04 2021 -0800

    Extract messages from the homepage (#774)

commit b91a5b0cedec7ee435f0d38952cd30cfcf5eb3be
Author: beets <beets@users.noreply.github.com>
Date:   Thu Jan 14 17:00:45 2021 -0800

    [i18n] Extract strings for place types (#773)

    - remove fallback to "Place" as a label for "Administrative Area x" or "Eurostat NUTS x"
    - remove pluralize
    - extract strings for all place types we display (or default to Place if unavailable)
    - fix subtitles for EN on place page, so we won't show e.g. "A Administrative Area of France"

    Note: I've decided to undo how we always show "Place" instead of Eurostat NUTS x or Administrative Area x. I've found it confusing.

    Pre:
    - Scotland: A Place in United Kingdom
    - Metropolis of Greater Paris: A Place in Île-de-France, France, Europe

    Post:
    - Scotland: Eurostat NUTS 1 Place in United Kingdom, Europe
    - Metropolis of Greater Paris: Administrative Area in Île-de-France, France, Europe

commit fca997b8b4fc2d98f93245ed64132b75ca02e768
Author: beets <beets@users.noreply.github.com>
Date:   Thu Jan 14 10:52:39 2021 -0800

    [i18n] Markup and extract strings for place ranking pages (#772)

    also adds a few straggler strings on the place pages

commit 8407a49f7d3c8596e0864ac598c99b5fe8bb7db1
Author: hanlu09205 <59888187+hanlu09205@users.noreply.github.com>
Date:   Tue Jan 12 15:11:12 2021 -0800

    All child places (#770)

    * translate allchildPlace & cache.memorize cached_i18n_name instead

commit 131864680177023b146957458d944ab8d24f9df2
Author: hanlu09205 <59888187+hanlu09205@users.noreply.github.com>
Date:   Mon Jan 11 20:55:28 2021 -0800

    Ranking i18n (#766)

    * adding more data plumbing

    * update comments

    * fix lint

    * revert change in types.ts

commit 318966faf8414447b93fdf5963eb1aa51c3a3edd
Author: beets <beets@users.noreply.github.com>
Date:   Mon Jan 11 13:54:54 2021 -0800

    Plumb through translations for more strings (#760)

    - Use translations for chart categories -- added the translations from the API call so that the category strings are sourced the same way. Note that the string "overview" was marked from translation from the client though, so there's a small hack in place.ts to use the client-side translation for that).
    - Use translation for related chart titles and added test
    - Fixed a couple of translations where the translators missed out a formatting character
    - Updated scripts to compile message bundle if required

commit d069616ca168e5bec6ac5a57bfc74f59efcf7298
Author: beets <beets@users.noreply.github.com>
Date:   Fri Jan 8 13:27:52 2021 -0800

    Import translated strings for place pages from TC (#758)

    - Adds translated strings for babel and formatjs
    - Removes old compiled babel bundles
    - Updates server and test scripts to compile bundles
    - Small update to tests

commit 60300c7d1aa139f73258201d4777341eb4674463
Author: hanlu09205 <59888187+hanlu09205@users.noreply.github.com>
Date:   Thu Jan 7 21:21:20 2021 -0800

    Pass locale to client (#756)

    * pass locale to client and then back

    * fix lint

    * update webdriver text

commit 3056418482bc952b5c8f37f30cfbe5700c903998
Author: hanlu09205 <59888187+hanlu09205@users.noreply.github.com>
Date:   Wed Jan 6 14:56:30 2021 -0800

    switch caller of get_display_name param to g.locale (#754)

    * pass in g.locale to get_display_name

commit a22f81de56e8db4a7e8d283421b631c2d321cc8c
Author: hanlu09205 <59888187+hanlu09205@users.noreply.github.com>
Date:   Wed Jan 6 11:25:49 2021 -0800

    Context locale 3 (#748)

    * Use request context to set locale to avoid plumbing through locale

    * use g.locale in api_display_name

    * update ranking.py call of get_i18n_name

commit f0a9ae5aa1a113a7d9de69fc2e574737f0db9f92
Author: beets <beets@users.noreply.github.com>
Date:   Tue Jan 5 17:29:08 2021 -0800

    [i18n] Merge in changes from master (#746)

    Co-authored-by: Eduardo Morales <edumorales@google.com>
    Co-authored-by: Bo Xu <shifucun@users.noreply.github.com>
    Co-authored-by: intrepiditee <31870027+intrepiditee@users.noreply.github.com>
    Co-authored-by: Prashanth R <shanth@google.com>
    Co-authored-by: hanlu09205 <59888187+hanlu09205@users.noreply.github.com>

commit ba82ce01b8dcc5ca1760b6388603cc7a811bfa0a
Author: beets <beets@users.noreply.github.com>
Date:   Tue Jan 5 17:07:57 2021 -0800

    Revert "Squashed commit of the following: (#741)" (#743)

    This reverts commit c87bf72.

    This last commit seems to have messed up future diffs on the i18n branch, possibly related to this:
    https://github.community/t/github-pull-requests-showing-invalid-diff-for-already-merged-branch/559

commit c87bf72de8f291dd5af7d1c77c05e60c2e91532f
Author: beets <beets@users.noreply.github.com>
Date:   Tue Jan 5 14:40:35 2021 -0800

    Squashed commit from master (#741)

commit 80ed172ebca99b08e4f941741563f86c83c8d2fc
Author: hanlu09205 <59888187+hanlu09205@users.noreply.github.com>
Date:   Tue Jan 5 14:15:54 2021 -0800

    Copy 736 to i18n (#739)

    * update landing page to respect locale param

    * revert logging change

    * update test_page_serve_usa title

    * copy https://github.com/datacommonsorg/website/pull/736/files to i18n

commit a522e29e408367d426520e4525213d9c6ad53d47
Author: hanlu09205 <59888187+hanlu09205@users.noreply.github.com>
Date:   Tue Jan 5 09:48:36 2021 -0800

    update landing page to respect hl param (#733)

    hl set to en:
    http://127.0.0.1:8080/place/geoId/06?hl=en
    https://screenshot.googleplex.com/BaCcD3zqFuqmS9S
    hl set to zh:
    http://127.0.0.1:8080/place/geoId/06?hl=zh
    https://screenshot.googleplex.com/3RjhrafQpyepL6y
    with topic set:
    http://127.0.0.1:8080/place/geoId/06?hl=zh&topic=Demographics
    https://screenshot.googleplex.com/3XBiTqrihZw3X7J

commit a802ce87e7e770a309de2c3203913b2aa07caa31
Author: hanlu09205 <59888187+hanlu09205@users.noreply.github.com>
Date:   Tue Jan 5 09:47:12 2021 -0800

    Switch to i18n name (#728)

    Make locale a default value in cached_i18n_name, get_i18n_name in case upstream missing it.
    Update cached_i18n_name to fall back to name, if there is no nameWithLanguage, which is legitimate for some places.
    Switch to use cached_i18n_name or get_i18n_name and have caller pass in locale.
    Added unit test

    Tested
    ranking.py
    http://127.0.0.1:8080/api/ranking/Count_Person/Place/geoId/06?hl=zh https://screenshot.googleplex.com/PA4xj69AY3R7b9p
    landing_page.py
    http://127.0.0.1:8080/api/landingpage/data/geoId/17043 https://screenshot.googleplex.com/5r7C3B2JavYMXDU
    chart.py http://127.0.0.1:8080/api/chart/geojson/geoId/06 https://screenshot.googleplex.com/Aa7CqX28LuiLAAj

commit af0cf189ba2309ceb0cf885f208e97fc022e798c
Author: hanlu09205 <59888187+hanlu09205@users.noreply.github.com>
Date:   Mon Dec 21 22:27:33 2020 -0800

    Fetchchild (#731)

    * add nameWithLanguage in api_ranking response

    * add test for no nameWithLanguages cases

commit 02876ad1c69582207bbed0a7a3e78f3246717b3f
Author: hanlu09205 <59888187+hanlu09205@users.noreply.github.com>
Date:   Fri Dec 11 20:29:19 2020 -0800

    api_i18n_name and corresponding functions, which we later could still… (#706)

    * api_i18n_name and corresponding functions, which we later could still name related functions to

    * Thanks for the review. Changed per comments.

    * update per comments.

    * fix sync problem

    * clean name_in_en

    * Remove website

    * Update README regarding bringing up local website

    * Small comment updates

    Apply suggestions from code review

    * update per comment

    * fix lint error.

    Co-authored-by: beets <beets@users.noreply.github.com>

commit 79adb217b0f794ec4eeead19b6fa6ac8d1ab98be
Author: beets <beets@users.noreply.github.com>
Date:   Fri Dec 11 14:32:27 2020 -0800

    Extract strings from place_landing.html (#712)

    Updated extract/update commands to use longer line widths
    Sorted the strings (and caught a few errors that way)
    Added placeholder es strings for testing

commit 89522eed8448598613fbf3a47de86a798339a19a
Author: beets <beets@users.noreply.github.com>
Date:   Thu Dec 10 15:15:11 2020 -0800

    Add chart topics to message bundle (#713)

    Found and fixed a few duplicate id bugs
    Added empty msgid/msgstr at the top of the file to avoid fuzzy errors during compilation

commit c315cf4606564f0b5dfe7d9c1fdb680122682b26
Author: beets <beets@users.noreply.github.com>
Date:   Wed Dec 9 14:43:28 2020 -0800

    Extract mesages from flask base template (#705)

    Also adds simple code to test using es translations (there are a few test strings in es/messages.po)

commit a61cb4021e593a4f46b6ca7c2b303b2a01c72f28
Author: beets <beets@users.noreply.github.com>
Date:   Fri Dec 4 13:18:23 2020 -0800

    Create a message json file for stat var labels (#697)

    Converted stats_var_label into a json file for TC extraction. A follow on will be to use these messages instead of the static array.

commit 5566192806f4c6604fbe7317993ac2f40dcd9ad8
Author: beets <beets@users.noreply.github.com>
Date:   Wed Dec 2 16:12:13 2020 -0800

    Squashed commit from master (#696)

commit e1e1ced09b28a155ed84654541762812edbc9e46
Author: beets <beets@users.noreply.github.com>
Date:   Fri Nov 20 09:38:54 2020 -0800

    Small updates to place UI strings (#684)

    * Update a few place UI strings and descriptions

    * fix a few more things

commit 0e6ab7302d5804f07f876bb7d8f03476f11fc025
Author: TJann <18621425+tjann@users.noreply.github.com>
Date:   Wed Nov 18 18:52:55 2020 -0800

    Localize place pages with FormatJS (#676)

commit 93f938e9b66a14aabe811d2032780746b0f9a59e
Author: beets <beets@users.noreply.github.com>
Date:   Wed Nov 18 10:05:19 2020 -0800

    Added descriptions for remaining charts (#679)

    * Added descriptions for remaining charts

    * Add a few more missing descriptions, update message description template

    * Update description for mortality events

commit f47c5c7f8f8cb90e00ad874263088c31bcab41f5
Author: beets <beets@users.noreply.github.com>
Date:   Mon Nov 16 14:34:30 2020 -0800

    Extract chart titles for translations (#677)

    - Added titleId and some chart descriptions to chart config
    - Adds a new script to extract titles from chart config
    - POC for using the translated string in the returned landing page response.

### CODE REVIEW

**Comment by beets** [comment]
Thankfully there weren't too many merge conflicts since there were no overlapping features developed at the same time. I took a detailed look at each file post-merge. It might be easier to see the diffs on github desktop than the web UI.

There are still a few more loose ends to tie up wrt i18n, but wanted to do the merge sooner so we can test on staging.

**Comment by shifucun** [comment]
wow, nice to see this land!

**Review by shifucun — APPROVED** [review]

**Comment by beets** [comment]
thanks! i'll close the i18n branch after this goes in

File: run_test.sh [hunk 1]
```diff
@@ -101,17 +105,15 @@ function run_py_test {
...
   cd server
   if [ ! -d dist  ]
   then
     echo "no dist folder, please run ./run_test.sh -b to build js first."
     exit 1
   fi
   export FLASK_ENV=webdriver
-  export GOOGLE_CLOUD_PROJECT=datcom-website-staging
-  pip3 install -r requirements.txt
+  export GOOGLE_CLOUD_PROJECT=datcom-browser-staging
```
Comment (shifucun) [inline run_test.sh:116]: datcom-website-staging

File: run_test.sh [hunk 2]
```diff
@@ -134,8 +135,7 @@ function run_screenshot_test {
   fi
 
   export FLASK_ENV=webdriver
-  export GOOGLE_CLOUD_PROJECT=datcom-website-staging
-  pip3 install -r requirements.txt
+  export GOOGLE_CLOUD_PROJECT=datcom-browser-staging
```
Comment (shifucun) [inline run_test.sh:138]: datcom-website-staging

File: run_test.sh (hunk 1 above)
Comment (beets) [reply run_test.sh]: thanks for the catch!

File: run_test.sh (hunk 2 above)
Comment (beets) [reply run_test.sh]: thanks for the catch!

File: server/routes/api/place.py [hunk 3]
```diff
@@ -432,14 +523,18 @@ def get_ranking_url(containing_dcid,
     return url
 
 
-@cache.memoize(timeout=3600 * 24)  # Cache for one day.
+@cache.cached(timeout=3600 * 24, query_string=True)  # Cache for one day.
```
Comment (shifucun) [inline server/routes/api/place.py:526]: i guess ideally we could use @cache.cached for all direct flask endpoints?

Comment (beets) [reply server/routes/api/place.py:526]: Hm, yes, especially now that there is an implicit get parameter that is used across endpoints.

**Pull request merged**